### DISCUSSION TITLE
Modernizing the codebase

### DISCUSSION THREAD

**Discussion by timofei-iatsenko** [lingui/js-lingui#2044] [General]
I will use this issue as an Epic for ongoing work for modernizing Lingui codebase.  

The goals of this Epic are:

- [x]  Migrate to ES Modules.
- [x] Migrate to Vite from Jest (#2039).
- [x] Update libraries which update was blocked by es modules.
- [x] Setup dependabot.

### Migrate to ES Modules

This is a challenging task because we are supporting babel macro plugin which [is not supporting ES modules yet](https://github.com/kentcdodds/babel-plugin-macros/issues/188), and this blocks all used in this plugin dependencies to be migrated to ESM. (CJS modules could be imported from ESM modules, but ESM modules could not be imported from CJS synchronously). I [started working](https://github.com/babel/babel/pull/16862) on ground changes in the babel to support ESM in the babel-plugin-macros

This would be a major update with a breaking changes, because everyone who uses lingui internals from CJS context will not be able to do so after switching to ESM. 

While we would be waiting for broader adoption of these changes in the babel & babel-plugin macros, I'm going to use `jiti` to provide CJS shim for ESM macro. 

Subtasks:

- Deprecate / remove pieces of code which could not be directly translated to ESM. [Here](https://github.com/lingui/js-lingui/blob/839f0b7927455e4598f00c62c3632c29dfd40dfe/packages/cli/src/api/extractors/index.ts#L25)
- Convert every package to `type: module`
- Drop CJS entries in packages
- Implement CJS shim for macro with jiti

### Migrate to Vite from Jest

Vite starts up much faster, needs less setup, and works well with ESM right away. While Jest has been popular for a long time, its many users make it hard to change things fundamentally without breaking existing setups. This has slowed down progress and made it tough to improve quickly.

**Comment by andrii-bodnar** [comment]
Thanks for bringing this up!

It sounds like a great initiative. However, I'd like to wrap up the v5 release first (most likely without the Vue integration since there's still a lot of work to be done there), and then we can plan these changes for Lingui v6.

Right now is a good time to deprecate things that don't work well in an ESM environment.
